feat(connlib): clients make use of DNS mangling on gateways

conectado opened the pull request:
This PR is the "client-side" of things for #4994. Up until now, when a user wanted to connect to a DNS resource, we would establish a connection to the gateway and pass along the domain we are trying to access. The gateway would resolve that domain and send the response back to the client, allowing them to finally send a DNS response.

Now, we instantly assign and respond with 4x A and 4x AAAA records to any query for one of our DNS resources. Upon the first IP packet for one of these "proxy IPs", we select a gateway, establish a connection and send our proxy IPs along. The gateway then performs the necessary mangling and NATing of all packets. See #5354 for details.

Resolves: #4994.
Resolves: #5491.

conectado committed:
refactor(connlib): assign ips that uniquely identifies an fqdn

This is not a working commit, though it does compiles.

With this refactor, each time a DNS query for one of our resources is
seen we return a unique set of ips for it. Might be the case we have
seen the resource before, that means we return forever the same one.

Now we don't initiate a connection when we see the query but we defer
that until a later time since mangling will be done in the gateway.

remove dns_internal_ips

send intent based on ip

refactor(connlib): remove translation on client peers

refactor(connlib): keep track of the fqdn and ips in the awaiting connection details and send that up to the gateway

chore(connlib): remove unused function

refactor(connlib): have the gateway translate from the proxy ips to the real ip addresses for DNS

refactor(connlib): do nat propperly with source port and translate destination ip

refactor(connlib): expire nat table after not seeing packet for 1 min

fix(connlib): delete the correct table entry when expirying nat

feat(connlib): add initial ipv4-to-ipv6 translation methods

feat(connlib): add ipv6 to ipv4 header translation for non-fragmented packets

chore(connlib): remove unused functions

fix(connlib): set translation ips in gateways for reuse connections

refactor(connlib): add padding for packets to be able to convert them

feat(connlib): implement ipv6 to ipv4 conversion with icmpv6 to icmpv4 still missing

fix(connlib): icmp conversion

feat(connlib): set ipv4 for ipv6 translation if ipv6 is empty and vice-versa

chore(connlib): split allowed_ips into ipv4 and ipv6 in `ClientOnGateway`

To encode into the types that clients always have both ipv4 and ipv6 and
they are the only allowed source ips for any given client, we split
those into their specific fields in the `ClientOnGateway` struct and
update tests accordingly.

Furthermore, these will be used for the DNS refactor for ipv6-in-ipv4
and ipv4-in-ipv6 to set the source ip of outgoing packets, without
having to do additional routing or mappings. There will be more notes on
this on the corresponding PR.

fix(connlib): don't shorten packets after decapsulation

fix(connlib): poll_read to account for additional 20 bytes

fix(connlib): payload_mut for convertable ip packets

fix(connlib): correctly set source instead of destination for translate_source

fix(connlib): ipv4 to ipv6 payload length calculation

fix(connlib): icmp echo and reply

fix(rust): to_owned for convertable packets

chore(connlib): add todo

chore(connlib): remove stray dbgs

fix(connlib): always ensure source before transforming packets

fix(connlib): obtain proxy ips

fix(connlib): send proxy ips for reusing resource

feat(connlib): refresh mapping after 30 seconds of no incoming traffic

fix semantic conflicts

fix(connlib) get proptests to compile

encode new test logic

fix(connlib): don't generate gaps for proxy-ips

remove unused dependency

fix(connlib): fix rust warnings

fix(connlib): clippy lints

rename DnsMap to StubDns

move dns_resources within stub_dns

rename peer transform functions to keep same scheme

fix typo

rename StubDns to StubResolver

simplify ip mapping

fix gateway compilation on windows

fix warning message

unify decapsulate functions in the gateway

unify decapsulate functions in the gateway

enable deserialization of old message format

add backward compatibility test

make gateway handle backward compatible messages

Fix compile errors

Fix itertools dep

Merge `pick_dns_resource` into `dns_resource_by_domain`

Use `sorted_by_key`

apply pr feedback

address PR feedback

simplify nat port assign

Construct `outside` only once

Use `inspect_err` instead of `match`

Update NAT logs

Only log if NAT entry was present

Fix typos

Fix warning

Minor updates to tests

Make errors around unsupported IP protocols more obvious

Rename functions for consistency

Ad some more docs

Add some basic proptests for NAT table

Fix test

thomaseizinger commented on rust/connlib/shared/src/messages.rs:
```diff
@@ -163,7 +163,7 @@ pub struct RequestConnection {
 pub struct ClientPayload {
     pub ice_parameters: Offer,
-    pub domain: Option<DomainName>,
+    pub domain: Option<(DomainName, Vec<IpAddr>)>,
```
In my tests I am using a `HashSet` here because there is no reason to allow duplicates and the order also shouldn't matter, right?

conectado replied:
I think order matter here for [round-robin](https://en.wikipedia.org/wiki/Round-robin_DNS), I try to preserve some order between the proxy-ip return order and the gateway assigned order.

thomaseizinger commented on rust/connlib/tunnel/src/gateway.rs:
```diff
@@ -167,6 +166,8 @@ impl GatewayState {
 
         let peer = self.peers.peer_by_ip_mut(dest)?;
 
+        let packet = peer.translate_real_to_proxy(packet, now)?;
```
Should we follow the same naming here with from/to network & TUN?

thomaseizinger commented on rust/connlib/tunnel/src/lib.rs:
```diff
@@ -64,15 +63,13 @@ pub struct Tunnel<CB: Callbacks, TRoleState> {
-    device_read_buf: Box<[u8; MTU]>,
+    write_buf: Box<[u8; MTU + 16 + 20]>,
+    // We have 20 extra bytes to be able to convert between ipv4 and ipv6
+    device_read_buf: Box<[u8; MTU + 20]>,
```
This `20` should probably be in a const somewhere.

conectado replied:
I'm thinking of improving this while thing, with some additional wrapper types around our buffers so I haven't cleaned this whole thing up yet

thomaseizinger commented on rust/connlib/tunnel/src/peer.rs:
```diff
@@ -176,6 +161,8 @@ impl ClientOnGateway {
             ipv6,
             resources: HashMap::new(),
             filters: IpNetworkTable::new(),
+            permanent_translations: Default::default(),
```
In case a client re-establishes their connection, we are okay to drop this state yeah? (Thinking ahead in how we are going to change roaming.)

conectado replied:
yeah :) it's fine because if the connection is being re-established the 5-tuple is already going to change killing any connection on top anyways

thomaseizinger commented on rust/connlib/tunnel/src/peer.rs:
```diff
@@ -186,6 +173,74 @@ impl ClientOnGateway {
+                                .map(|ip| ip.network_address())
+                                .cycle(),
+                        )
+                    };
```
We can probably move some of these onto `ResourceDescription` to make this a bit easier to read. A comment on what we are doing here also wouldn't hurt.

thomaseizinger commented on rust/connlib/tunnel/src/peer.rs:
```diff
@@ -186,6 +173,74 @@ impl ClientOnGateway {
+                for (proxy_ip, real_address) in resource_ips
+                    .iter()
+                    .filter(|ip| ip.is_ipv4())
+                    .zip(real_ipv4_mappings)
```
Perhaps we should call `.cycle` here. After all, I am guessing the requirement comes from the fact that `zip` would bail early if the 2nd iterator isn't long enough?

conectado replied:
the problem with moving the `cycle` here is that now the boxed iterator needs to implement clone, so we need to create a super trait to express that. Which we can do but I think that's a bit harder to read than this?

conectado replied:
I think I have a fix for both this and this https://github.com/firezone/firezone/pull/5049#discussion_r1623869742

conectado replied:
just running the tests to make sure I didn't mess it up

conectado replied:
https://github.com/firezone/firezone/pull/5049/commits/a832ee98990abaa22df3a5ae88a874ab7226447a

thomaseizinger commented on rust/connlib/tunnel/src/dns.rs:
```diff
@@ -97,76 +78,126 @@ impl<T, V> ResolveStrategy<T, DnsQueryParams, V> {
+    fn get_records<N: ToName>(&mut self, question: &Question<N>) -> Vec<RecordData<DomainName>> {
+        match question.qtype() {
+            Rtype::A => self
+                .get_or_assign_ips(question.qname().to_name())
```
As discussed. I think we should lazily allocate IPs and not assign IPv6 IPs if we are only asked for IPv4.

conectado replied:
I have to refactor how peer selection/connection work, will rework that then

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -317,20 +265,13 @@ pub struct ClientState {
     ///
     /// "Refreshing" DNS resources means triggering a new DNS lookup for this domain on the gateway.
     next_dns_refresh: Option<Instant>,
+    dns_map: DnsMap,
```
I'd prefer if we can adopt a naming here that follows what DNS itself uses. Technically speaking, I think we are building an "(authoritative) name server" here?

conectado replied:
`stub_resolver` might be a better name

thomaseizinger replied:
Yeah true!

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -614,84 +548,21 @@ impl ClientState {
-        now: Instant,
     ) -> Result<Option<IpPacket<'a>>, (MutableIpPacket<'a>, IpAddr)> {
-        match dns::parse(
+        match self.dns_map.parse(
```
I think it is rather odd that we are calling `parse` here on this now stateful component. I think a name like `handle` would fit better.

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -614,84 +548,21 @@ impl ClientState {
         &mut self,
         packet: MutableIpPacket<'a>,
-        now: Instant,
     ) -> Result<Option<IpPacket<'a>>, (MutableIpPacket<'a>, IpAddr)> {
```
The return type here is dated too, the caller should just call `.destination` IMO.

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -1139,6 +936,24 @@ impl ClientState {
+            .collect_vec(),
+        ResourceDescription::Cidr(r) => vec![r.address],
+    }
+}
```
Move this onto `AwaitingConnectionDetails`?

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -1272,6 +1087,30 @@ impl IpProvider {
+            }
+        }
+
+        res
```
Isn't this what the `take()` combinator does?

thomaseizinger reviewed (commented):
Did an early pass. Looks pretty good already! Curious to hear whether my new tests found any bugs!

thomaseizinger commented on rust/connlib/tunnel/src/tests.rs:
```diff
@@ -2595,6 +2615,37 @@ fn assert_dns_packets_properties(state: &TunnelTest, ref_state: &ReferenceState)
+        }
+        _ => false,
+    }
+}
```
There is already a utility called `matches_domain` that should help you here!

conectado replied:
`matches_domain` is a bit more limited, since it matches something like `?.foo` to `bar.baz.foo`

conectado replied:
maybe we should replace `matches_domain` with this?

thomaseizinger replied:
> maybe we should replace `matches_domain` with this?

Yeah. Can we implement it in a simpler way? Or with less duplication?

conectado replied:
maybe it's better to use `is_subdomain` that's implemented in the dns module? it's quite simple and it has some separated unit-tests so it's not that circular

thomaseizinger replied:
Hmm, I guess it is okay? I didn't really want to depend on implementation details like this in the test too but if it is separately tested should be fine.

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -332,24 +280,15 @@ pub struct ClientState {
     mangled_dns_queries: HashMap<u16, std::time::Instant>,
     /// When to next refresh DNS resources.
     ///
     /// "Refreshing" DNS resources means triggering a new DNS lookup for this domain on the gateway.
```
Outdated docs.

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -465,22 +425,23 @@ impl ClientState {
             now,
         );
 
+        if peer.allowed_ips.longest_match(dest).is_none() {
+            let gateway_id = peer.id();
+            self.send_proxy_ips(&dest, resource, gateway_id);
+            return None;
+        }
```
Can you add a comment explaining why we are doing this?

conectado replied:
yeah, the quick and dirty is that we might have a connection with the desired gateway but we haven't send the ips yet. I'll write a comment when I generating the ipv4/ipv6 proxy ips since this might need a small tweak :)

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -518,7 +479,7 @@ impl ClientState {
         };
 
-        let packet = match peer.transform_network_to_tun(packet) {
+        let packet = match peer.ensure_allowed(packet) {
```
`ensure_allowed` returning a new packet is kind of odd?

conectado replied:
good point, just renamed the old function because we no longer do any transformation here, will change this to an `&mut` that returns a bool

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -565,24 +525,23 @@ impl ClientState {
+            return Err(Error::ControlProtocolError);
+        };
+
+        // TODO: can we create this on connection intent now?
```
Either address or delete the comment :)

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -959,42 +813,20 @@ impl ClientState {
         self.node.handle_timeout(now);
 
-        match self.next_dns_refresh {
+        match self.next_mangled_dns_expire {
```
Would you mind sending a separate PR for this?

thomaseizinger commented on rust/connlib/tunnel/src/dns.rs:
```diff
@@ -27,13 +27,11 @@ const DNS_PORT: u16 = 53;
 #[derive(Debug)]
-pub(crate) enum ResolveStrategy<T, U, V> {
+pub(crate) enum ResolveStrategy<'a> {
     /// The query is for a Resource, we have an IP mapped already, and we can respond instantly
```
Slightly outdated comment, the "instantly" is now obsolete because we don't defer DNS responses anymore.

thomaseizinger commented on rust/connlib/tunnel/src/dns.rs:
```diff
@@ -108,76 +88,128 @@ impl<T, V> ResolveStrategy<T, DnsQueryParams, V> {
+            .fqdn_to_ips
+            .entry(fqdn.clone())
+            .or_insert_with(|| {
+                // TODO: the side effeccts are executed even if this is not inserted
+                // make it so that's not the case
```
Outdated comment.

thomaseizinger commented on rust/connlib/tunnel/src/dns.rs:
```diff
@@ -220,8 +252,11 @@ fn build_response(original_pkt: IpPacket<'_>, mut dns_answer: Vec<u8>) -> IpPack
 
+    // TODO: this is some weirdness due to how MutableIpPacket is implemented
+    // we need an extra 20 bytes padding.
+    res_buf.extend_from_slice(&[0; 20]);
```
I've always found this function to be very odd. Why are operating on byte buffers? Can we use `ip_packet::make` here instead now?

thomaseizinger commented on rust/connlib/tunnel/src/dns.rs:
```diff
@@ -261,33 +298,32 @@ where
     );
 
-    let Some(resource) = resource else {
+    if records.is_empty() {
```
I've looked this up recently and no records doesn't necessarily mean NXDOMAIN. It is a valid DNS answer to return an empty set of records if the name is known but doesn't have any records attached.

I don't think there is a case where connlib should make an NXDOMAIN response:

- If the query is for a resource, we know the name.
- If we don't know the name, we need to forward and thus the next resolver needs to handle this.

thomaseizinger commented on rust/connlib/tunnel/src/gateway.rs:
```diff
@@ -197,6 +209,16 @@ impl GatewayState {
+
+        let packet = peer.translate_proxy_to_real(packet, now)?;
+
         if let Err(e) = peer.ensure_allowed(&packet) {
```
We are calling 3 functions on `peer` here. Should we introduce a `peer.decapsulate`? (or `translate` / `transform`?)

thomaseizinger commented on rust/connlib/tunnel/src/gateway.rs:
```diff
@@ -229,14 +251,14 @@ impl GatewayState {
         stun_servers: HashSet<SocketAddr>,
         turn_servers: HashSet<(RelayId, RelaySocket, String, String, String)>,
-        domain: Option<DomainName>,
+        domain: Option<(DomainName, Vec<IpAddr>)>,
```
Should we make a struct for this tuple?

thomaseizinger commented on rust/connlib/tunnel/src/gateway.rs:
```diff
@@ -229,14 +251,14 @@ impl GatewayState {
-            (Some(domain), ResourceDescription::Dns(r)) => {
-                if !crate::dns::is_subdomain(domain, &r.domain) {
+            (Some((domain, addresses)), ResourceDescription::Dns(r)) => {
+                if !crate::dns::is_subdomain(domain, &r.domain) || addresses.is_empty() {
```
So this means, a DNS resource that doesn't have any records yet is just invalid? I think we should be careful with failing from the `accept` function because the client has no idea what is happening then and will just time-out, making this rather hard to debug.

At the minimum, `addresses.is_empty` should log a fat warning on the gateway.

Why do we need to fail the connection here just because there are no DNS records?

thomaseizinger replied:
This also won't work with our idea of making things backwards compatible.

thomaseizinger commented on rust/connlib/tunnel/src/gateway.rs:
```diff
@@ -336,9 +354,10 @@ impl GatewayState {
+                self.peers.iter_mut().for_each(|p| {
+                    p.expire_resources(utc_now);
+                    p.expire_nat(now)
+                });
```
Eventually, it would be good to continue the model of `handle_timeout` all the way to the `ClientOnGateway` struct.

thomaseizinger commented on rust/connlib/tunnel/src/gateway.rs:
```diff
@@ -382,7 +401,17 @@ impl GatewayState {
+            if let Some(ev) = peer.poll_event() {
+                return Some(ev);
+            }
+        }
```
In `snownet`, I am doing this kind of work in `handle_timeout` and just move all the events in `buffered_events`, you can do something like:

```rust
self.buffered_events.extend(self.peers.iter_mut().flat_map(|peer| iter::from_fn(|| peer.poll_event())));
```

thomaseizinger replied:
It is better to treat `handle_timeout` as the kind of "work-horse" of these state machines and keep getters like `poll_event` simple.

conectado replied:
so you move the `buffered_events` in the state `handle_timeout`?

thomaseizinger commented on rust/connlib/tunnel/src/peer.rs:
```diff
@@ -167,6 +153,106 @@ impl ClientOnGateway {
         [IpAddr::from(self.ipv4), IpAddr::from(self.ipv6)]
     }
 
+    pub(crate) fn refresh_translation(
```
If you model this component with `handle_timeout`, you might be able to write a much simpler `proptest-state-machine` test that takes into account time as well and ensures that the NAT table expires correctly.

The difficulty with time-based tests is correctly modelling, what expires when. But in the scope of just the NAT table, this might be easy enough?

thomaseizinger commented on rust/gateway/src/eventloop.rs:
```diff
@@ -110,6 +116,22 @@ impl Eventloop {
+                    )
+                    .is_err()
+                {
+                    tracing::warn!("Too many connections requests, dropping existing one");
```
Copy-pasta error.

thomaseizinger commented on rust/gateway/src/eventloop.rs:
```diff
@@ -251,33 +283,36 @@ impl Eventloop {
+    pub fn allow_access(&mut self, result: Result<Vec<IpAddr>, Timeout>, req: AllowAccess) {
         let addresses = result
             .inspect_err(|e| tracing::debug!(client = %req.client_id, reference = %req.reference, "DNS resolution timed out as part of allow access request: {e}"))
             .unwrap_or_default();
```
I think this `unwrap_or_default` will also not play well with your `is_empty` check inside `allow_access`.

thomaseizinger commented on rust/gateway/src/eventloop.rs:
```diff
@@ -305,7 +340,7 @@ fn resolve_addresses(_: &str) -> std::io::Result<Vec<IpNetwork>> {
 
 #[cfg(not(target_os = "windows"))]
-fn resolve_addresses(addr: &str) -> std::io::Result<Vec<IpNetwork>> {
+fn resolve_addresses(addr: &str) -> std::io::Result<Vec<IpAddr>> {
```
You also need to change the signature for windows. It appears that we are not checking in CI that it at least compiles on Windows?

thomaseizinger commented on rust/ip-packet/src/lib.rs:
Shall we extract this into a separate PR and add some basic tests that convert back and forth between a few packets?

conectado replied:
yeah :)

thomaseizinger reviewed (commented):
Did another pass on this! Great work so far :)

Most of the comments are just suggestions!

thomaseizinger commented on rust/connlib/tunnel/src/dns.rs:
```diff
@@ -108,76 +90,129 @@ impl<T, V> ResolveStrategy<T, DnsQueryParams, V> {
+    // TODO: we can save a few allocations here still
+    pub(crate) fn handle<'a>(
+        &mut self,
+        dns_mapping: &bimap::BiMap<IpAddr, DnsServer>,
```
Did you consider moving this into the state of the resolver too? I think that makes sense. Those are the servers the resolver forwards to if it can't handle a query.

Can wait for a follow-up though :)

conectado replied:
yeah, let's do it in a follow up PR

conectado committed:
refactor(gateway): support translation of dns

thomaseizinger committed:
Fix bad index into buffer

thomaseizinger committed:
Make `ensure_allowed` on client match gateway-side design

thomaseizinger committed:
Make `handle_timeout` pure

thomaseizinger committed:
Remove `DNS_REFRESH_INTERVAL` const

thomaseizinger committed:
Remove `next_mangled_dns_expire` field

thomaseizinger committed:
Align test implementation of DNS resources with CIDR resources

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -621,24 +550,28 @@ impl ClientState {
             })));
         };
 
+        self.peers.insert(
+            GatewayOnClient::new(gateway_id, &ips, HashSet::from([resource_id])),
+            &[],
+        );
```
@conectado Because we need access to `awaiting_connection_details` (which we remove in this function now) in order to create the `GatewayOnClient` in `peers`, I had to move the creation of the peer into this function.

thomaseizinger committed:
Remove duplicate log about DNS server as CIDR resource

thomaseizinger committed:
Don't expire queries early

Fixes: #5491.

thomaseizinger committed:
Don't double check for expired queries

Very likely, expired queries will no longer be present in the map by the time we get the response because we clear them in `handle_timeout`.

thomaseizinger committed:
Log domains and rtt for DNS query
